Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionTreeViewItems.cs

```diff
@@ -0,0 +1,530 @@
+
+        // If we're a singleton action (no associated action map property), we include all our bindings in the
+        // serialized data.
+        public override bool serializedDataIncludesChildren => actionMapProperty == null;
+
+        public override void Rename(string newName)
+        {
+            InputActionSerializationHelpers.RenameAction(property, actionMapProperty, newName);
+        }
+
+        public override void DeleteData()
+        {
+            InputActionSerializationHelpers.DeleteActionAndBindings(actionMapProperty, guid);
+        }
+
+        public override bool AcceptsDrop(ActionTreeItemBase item)
+        {
+            return item is BindingTreeItem && !(item is PartOfCompositeBindingTreeItem);
+        }
+
+        public override bool GetDropLocation(Type itemType, int? childIndex, ref SerializedProperty array, ref int arrayIndex)
+        {
+            // Drop bindings into binding array.
+            if (typeof(BindingTreeItem).IsAssignableFrom(itemType))
+            {
+                array = bindingsArrayProperty;
+
+                // Indexing by tree items is relative to each action but indexing in
+                // binding array is global for all actions in a map. Adjust index accordingly.
+                // NOTE: Bindings for any one action need not be stored contiguously in the binding array
+                //       so we can't just add something to the index of the first binding to the action.
+                arrayIndex =
+                    InputActionSerializationHelpers.ConvertBindingIndexOnActionToBindingIndexInArray(
+                        array, name, childIndex ?? -1);
+
+                return true;
+            }
+
+            // Drop other actions next to us.
+            if (itemType == typeof(ActionTreeItem))
+            {
+                array = arrayProperty;
+                arrayIndex = this.arrayIndex + 1;
+                return true;
+            }
+
+            return false;
+        }
+
+        public static ActionTreeItem AddTo(TreeViewItem parent, SerializedProperty actionMapProperty, SerializedProperty actionProperty)
+        {
+            var item = new ActionTreeItem(actionMapProperty, actionProperty);
+
+            item.depth = parent.depth + 1;
+            item.displayName = item.name;
+            parent.AddChild(item);
+
+            return item;
+        }
+
+        /// <summary>
+        /// Add items for the bindings of just this action to the given parent tree item.
+        /// </summary>
+        public void AddBindingsTo(TreeViewItem parent)
+        {
+            var isSingleton = actionMapProperty == null;
+            var bindingsArrayProperty = isSingleton
+                ? property.FindPropertyRelative("m_SingletonActionBindings")
+                : actionMapProperty.FindPropertyRelative("m_Bindings");
+
+            var bindingsCountInMap = bindingsArrayProperty.arraySize;
+            var bindingCountForAction = 0;
```

Comment:
new warning warning CS0219: The variable 'bindingCountForAction' is assigned but its value is never used